### PR TITLE
buckets and splitBy operators

### PR DESCRIPTION
**Pull request by backbone87**
- [x] Add the operator to Rx
- [x] It must have a `-spec.ts` tests file covering the canonical corner cases, with marble diagram tests
- [x] The spec file should have a type definition test at the end of the spec to verify type definition for various use cases -- *is the type test in splitBy-spec sufficient?*
- [ ] The operator must be documented in JSDoc style in the implementation file, including also the PNG marble diagram image -- *how to make the marble diagram?*
- [x] The operator should be listed in `docs_app/content/guide/operators.md` in a category of operators
- [x] The operator should also be documented. See [Documentation Guidelines](../CONTRIBUTING.md).

**Description:**

Introduces 2 new operators `buckets` and `splitBy`.

`buckets` uses a hash function to partition values across multiple multicasted (hot) observables. This behaves similar to `groupBy`, but the groups are known immediately.

`splitBy` is the special case of `buckets` where the count of buckets is 2. This is similar to `partition`, but has a usage ergonomy more similar to `groupBy`.

implementation is based upon `bucketBy` and `splitBy` in `rxjs-etc` by @cartant 

**Related issue (if exists):** https://github.com/ReactiveX/rxjs/issues/3807 https://github.com/ReactiveX/rxjs/issues/4419 https://github.com/ReactiveX/rxjs/issues/5731

### CODE REVIEW

File: spec/operators/splitBy-spec.ts [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+      expectObservable(split).toBe(expected, { x, y });
+    });
+  });
+
+  it('should infer partition types when using a type guard predicate', () => {
+    of(5, 'a', []).pipe(
+      splitBy((value): value is number => typeof value === 'number'),
+      mergeMap(([numbers, strings]) => merge(numbers.pipe(map((value) => value * value)), strings.pipe(map((value) => value.length))))
+    );
+  });
```
Comment (cartant) [inline spec/operators/splitBy-spec.ts:34]: We favour using dtslint for type-related tests:

https://github.com/ReactiveX/rxjs/blob/c23e567ccaee4a9aa2a7046519e7cf46a7191a3f/spec-dtslint/operators/filter-spec.ts#L37-L40

The tests are never executed, instead there is a lint rule that runs evaluates the types and dtslint expectations. To run them, use `npm run dtslint`.

File: src/internal/operators/buckets.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+import { operate } from '../util/lift';
+import { OperatorSubscriber } from './OperatorSubscriber';
+
+export interface BucketsOptions<T> {
+  /**
+   * The hash function used to partition the values into their buckets. The
+   * default hash function converts the values into numbers, if they are not
+   * already.
+   */
+  hashFn?: (value: T) => number;
```
Comment (cartant) [inline src/internal/operators/buckets.ts:13]: Personally, I'm not a fan of abbreviations. I'd prefer `hasher`, instead, but this can be bikeshedded later.

File: src/internal/operators/splitBy.ts [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+ *  containing the two observables for matching and non-matching values
+ *
+ * @see {@link buckets}
+ * @see {@link groupBy}
+ * @see {@link partition}
+ */
+export function splitBy<T>(
+  predicate: (value: T) => boolean,
+  options: SplitByOptions<T> = {}
+): OperatorFunction<T, [Observable<T>, Observable<T>]> {
```
Comment (cartant) [inline src/internal/operators/splitBy.ts:92]: This function has overload signatures and they hide the implementation signature. So you'll need this to be an overload signature in its own right - even though the implementation is exactly the same. That's my understanding, at least.

**Review by cartant — COMMENTED** [review]
Some initial comments. Will review it properly, later. Also, for PRs that change the public API, you will need to use `npm run api_guardian:update`.

File: src/internal/operators/buckets.ts (hunk 2 above)
Comment (backbone87) [reply src/internal/operators/buckets.ts:13]: i also just picked this up from somewhere, bikeshed away ;)

File: src/internal/operators/splitBy.ts (hunk 3 above)
Comment (backbone87) [reply src/internal/operators/splitBy.ts:92]: actually the options are just missing from the other signatures. typescript only shows the overloads to the callers not the impl signature. the only requirement for the impl signature is that is has to be assignable to each overload signature.

will fix the missing argument

**Comment by backbone87** [comment]
in regards to naming: is `intoBuckets` a better choice?

**Comment by cartant** [comment]
I think it could probably be a single operator - `splitBy` - with two signatures with different option types - `SplitByPredicateOptions` and `SplitByHashOptions` - with the predicate and count parameters moved into the respective options types. IDK. Much bikeshedding is possible.

**Comment by backbone87** [comment]
personally, i dont like merging them into a single operator. too overloaded from user & dev perspective. but that's just my feeling. `splitInto(3)`, `splitBy(({ name }) => name === 'rxjs')` just reads/sounds nice (not a native speaker though)

**Comment by cartant** [comment]
TBH, I was not planning on bringing `bucketBy` - or whatever - into this codebase. I was only going to bring in `splitBy`. I don't anticipate there being much appetite for an additional operator. Especially one that hasn't been requested. Expect this to be bikeshedded, too.

**Comment by benlesh** [comment]
So looking over this, at first blush. `buckets` looks like a specialized `groupBy`. and `splitBy` is a specialized `buckets`. So it makes me wonder if we need either of these. 🤔  Just thinking out loud here.

```ts
const buckets = (count) => groupBy((value, i) => i % count)
// or
const buckets = ({ hasher }) => groupBy(hasher);
// or
const splitBy = (predicate) => groupBy(predicate);
```

Am I off-base? What am I missing?

**Comment by benlesh** [comment]
To be clear @backbone87 ... this is awesome work! And I really appreciate it. However, we're going to be on the hook to support this API long-term, so I want to be sure we're doing the right thing.

**Comment by benlesh** [comment]
Another thing to consider... And `groupBy` handles this... but there's a scenario with operators like this, where the consumer can take the observable out of the result and subscribe to it, then early-unsub from the source. In those cases, the expectation would be that the created "buckets" or "groups" would still work until they were unsubbed. At least that's true for `groupBy`, so I guess I would expect the same behavior here.

```ts
const source = interval(1000);
const result = source.pipe(buckets(3));

const capturedBucketSubs = [];

const outerSubs = result.subscribe(bucket => {
   capturedBucketSubs.push(capturedBucketSubs.subscribe(console.log));
});

// unsubbing outer first shouldn't kill existing inners
setTimeout(() => {
  outerSubs.unsubscribe();
}, 5000);

setTimeout(() => {
  for (const subs of capturedBucketSubs) {
    subs.unsubscribe();
  }
}, 10000);
```

I'm willing to agree this behavior is SUPER debatable, but there it is, and that's how groupBy has worked for years, for better or worse.

**Comment by benlesh** [comment]
Sorry, too, @cartant, if I'm just springing this on you. When we talked about this at the core team meeting, I guess I was thinking `splitBy` did something different, but I'm not sure now, I guess, what I thought it did that `groupBy` couldn't do.

**Comment by backbone87** [comment]
`buckets` is not a special case of `groupBy`. check the examples in the doc comments which demonstrate the difference in behavior. the buckets are also subjects, so unsubbing from them should not change anything, but i can add a test for this.

**Comment by cartant** [comment]
@benlesh As I mentioned above, I was planning on ignoring `bucketBy`/`buckets` when I eventually got around to dealing with #3807. I don't think it adds anything that's particularly useful. It's just how `splitBy` was implemented in `rxjs-etc`. And adding random operators to that package does not come with the implications of adding operators to the core.

FWIW, I kinda agree with you on dumping this in favour of `groupBy` and a `partition` deprecation that links to a `deprecations/partition.md` document that explains how to use `groupBy` instead.

When I added `splitBy` to `rxjs-etc`, `groupBy` didn't have the typings that it does now. 12 months or so ago [Moshe added typings that made it play nice with user-defined type guards](https://github.com/ReactiveX/rxjs/pull/5441), so it composes pretty nicely:

```ts
dogsAndCats.pipe(
  groupBy(isDog),
  mergeMap(grouped => grouped.key
    ? /* Observable<Dog> */ grouped.pipe(tap(() => console.log('woof')))
    : /* Observable<Cat> */ grouped.pipe(tap(() => console.log('meow')))
)
```

I don't know that there is any advantage in using `splitBy` instead and being presented with the two observables up front - this is the difference mentioned in [the original issue](https://github.com/ReactiveX/rxjs/issues/3807):

> This is conceptually similar to groupBy with only two groups, except it returns an Array of the grouped Observables instead of an Observable of the groups since the groups can be known ahead of time.

Rather, I think `splitBy` will be more difficult for peeps to grok, as the two known-ahead-of-time observables will need to be combined into a single observable:

```ts
dogsAndCats.pipe(
  groupBy(isDog),
  mergeMap(([dog$, cat$])=> merge(
    dog$.pipe(tap(() => console.log('woof'))),
    cat$.pipe(tap(() => console.log('meow')))
  )
)
```

**TL/DR: I don't see a compelling case for `splitBy` either, so close this - and the related issues - and document the `partition` deprecation with examples on the docs site.**

**Comment by benlesh** [comment]
@cartant maybe if we get everything in a monorepo, `rxjs-etc` might be a solid candidate to join this repository, so we can direct things over there? Obviously debatable, but it might open the door for moving lesser used core operators over, as well. Just something to think about.

**Comment by backbone87** [comment]
no hard feelings here on my side, a little bit of disappointment, but I can handle this, if it doesn't fit, then leave this out.

personally, I have used partition only on very few occasions, but when I did, I definitely wasn't happy with its behavior as it is right now and would have preferred the splitBy behavior. also, these situations could have most likely also been solved with groupBy.

**Comment by benlesh** [comment]
After discussing this PR in the Core Team meeting, I'm so sorry to say that I don't think we're going to go ahead with it at this time, and we should probably close the related issue. That issue, TMK, was filed at a time when `groupBy` was a little less capable than it is now. While `splitBy` _is indeed_ a little different, we've decided that the difference isn't sufficient enough to justify supporting the API as part of the library for years to come.

Thank you so much for your hard work, @backbone87, we all really do appreciate it. I'm sorry that this isn't something we can merge into core for now. Especially given that it was just trying to implement something that was requested by a core team member years ago.

Thank you again, @backbone87, and I hope this doesn't put you off from further contributions.

**Comment by backbone87** [comment]
hm ok, no problem. i suspected it.
i have a few questions:
- what is the way forward with `partition`? the operator is already deprecated, but the behavior of the factory is still to consume twice instead of multicasting
- is there a way to achieve the bucket/splitBy behavior with another operator or pipeline of operators?
- would it be an alternative to add options to groupBy to enable a similar behavior to bucket/splitBy. something like `groupBy(({ key }) => key, { seed: () => of('key1', 'key2') })` where `seed` emits keys for which groups are eagerly opened and when it completes the `groupBy` completes to downstream, but still receives values from upstream and routes them into existing groups. unrouteable upstream values will trigger a stopped notification hook and `seed` defaults to `() => NEVER`, which means the `groupBy` does not create any groups eagerly and completes with the source.

**Comment by Polyterative** [comment]
I'm also interested in a way to replicate the `splitBy` with classic operators
